pkg/k8s/operations.go

```diff
@@ -333,6 +344,148 @@ func ApplyYaml(params ApplyYamlPrams, kubeconfig string, isLocal bool) (output s
 	return outStr, nil
 }
 
+// UpgradeInfra upgrades the Chaos Infrastructure using the provided manifest and kubeconfig with the help of client-go library.
+
+func UpgradeInfra(manifest []byte, kubeconfig string) (string, error) {
+
+	// Get Kubernetes and dynamic clients along with the configuration.
+
+	_, kubeClient, dynamicClient, err := getClientAndConfig(kubeconfig)
+	if err != nil {
+		return "", err
+	}
+
+	// Decode the manifest into a list of Unstructured resources.
+
+	resources, err := decodeManifest(manifest)
+	if err != nil {
+		return "", err
+	}
+
+	// Apply the decoded resources using the dynamic client and Kubernetes client.
+
+	err = applyResources(resources, dynamicClient, kubeClient)
+	if err != nil {
+		return "", err
+	}
+
+	logrus.Println("🚀 Successfully Upgraded Chaos Infra")
+
+	return "Success", nil
+
+}
+
+// retrieves the Kubernetes and dynamic clients along with the configuration.
+
+func getClientAndConfig(kubeconfig string) (*rest.Config, *kubernetes.Clientset, dynamic.Interface, error) {
+	var config *rest.Config
+	var dynamicClient dynamic.Interface
+
+	// If kubeconfig is provided, use it to create the configuration and dynamic client.
+
+	if kubeconfig != "" {
+		var err error
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to build config from flags: %w", err)
+		}
+		dynamicClient, err = dynamic.NewForConfig(config)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to create dynamic client: %w", err)
+		}
+	} else {
+
+		// Use the default kubeconfig file at $HOME/.kube/config.
+
+		home := homedir.HomeDir()
+		defaultKubeconfig := filepath.Join(home, ".kube", "config")
+		if _, err := os.Stat(defaultKubeconfig); !os.IsNotExist(err) {
+			var err error
+			config, err = clientcmd.BuildConfigFromFlags("", defaultKubeconfig)
+			if err != nil {
+				return nil, nil, nil, fmt.Errorf("failed to build config from flags: %w", err)
+			}
+			dynamicClient, err = dynamic.NewForConfig(config)
+			if err != nil {
+				return nil, nil, nil, fmt.Errorf("failed to create dynamic client: %w", err)
+			}
+		} else {
+			// Return an error if kubeconfig is not provided and the default file doesn't exist
+			return nil, nil, nil, errors.New("kubeconfig not provided, and default kubeconfig not found")
+		}
+	}
+
```

Comment:
This will reduce the repetition of the code for creating the dynamicClient

```suggestion
	if kubeconfig != "" {
		var err error
		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
		if err != nil {
			return nil, nil, nil, fmt.Errorf("failed to build config from flags: %w", err)
		}
	} else {
		// Use the default kubeconfig file at $HOME/.kube/config.
		home := homedir.HomeDir()
		defaultKubeconfig := filepath.Join(home, ".kube", "config")
		if _, err := os.Stat(defaultKubeconfig); !os.IsNotExist(err) {
			var err error
			config, err = clientcmd.BuildConfigFromFlags("", defaultKubeconfig)
			if err != nil {
				return nil, nil, nil, fmt.Errorf("failed to build config from flags: %w", err)
			}
		} else {
			// Return an error if kubeconfig is not provided and the default file doesn't exist
			return nil, nil, nil, errors.New("kubeconfig not provided, and default kubeconfig not found")
		}
	}
	dynamicClient, err = dynamic.NewForConfig(config)
	if err != nil {
		return nil, nil, nil, fmt.Errorf("failed to create dynamic client: %w", err)
	}
```